<!DOCTYPE html>













<html class="theme-next gemini" lang="en">
<head><meta name="generator" content="Hexo 3.8.0">
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
























<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2">

<link rel="stylesheet" href="/css/main.css?v=7.1.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=7.1.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=7.1.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=7.1.0">


  <link rel="mask-icon" href="/images/logo.svg?v=7.1.0" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Gemini',
    version: '7.1.0',
    sidebar: {"position":"left","display":"post","offset":12,"onmobile":false,"dimmer":false},
    back2top: true,
    back2top_sidebar: false,
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":false,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="First install the required packages data.table, Matrix, and SingleCellExperiment.1234567891011121314if (!requireNamespace(&quot;data.table&quot;, quietly = TRUE)) &amp;#123;    install.packages(&quot;data.table&quot;)&amp;#125;i">
<meta name="keywords" content="Single Cell,Script,Cell Ranger,R">
<meta property="og:type" content="article">
<meta property="og:title" content="An R script to process Cell Ranger output files">
<meta property="og:url" content="https://zhewa.github.io/2019/05/01/processCellRangerOutputs/index.html">
<meta property="og:site_name" content="singlecell.me">
<meta property="og:description" content="First install the required packages data.table, Matrix, and SingleCellExperiment.1234567891011121314if (!requireNamespace(&quot;data.table&quot;, quietly = TRUE)) &amp;#123;    install.packages(&quot;data.table&quot;)&amp;#125;i">
<meta property="og:locale" content="en">
<meta property="og:updated_time" content="2019-05-05T16:27:08.542Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="An R script to process Cell Ranger output files">
<meta name="twitter:description" content="First install the required packages data.table, Matrix, and SingleCellExperiment.1234567891011121314if (!requireNamespace(&quot;data.table&quot;, quietly = TRUE)) &amp;#123;    install.packages(&quot;data.table&quot;)&amp;#125;i">



  <link rel="alternate" href="/atom.xml" title="singlecell.me" type="application/atom+xml">



  
  
  <link rel="canonical" href="https://zhewa.github.io/2019/05/01/processCellRangerOutputs/">



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>An R script to process Cell Ranger output files | singlecell.me</title>
  




  <script async src="//www.googletagmanager.com/gtag/js?id=UA-138890986-1"></script>
  <script>
    var host = window.location.hostname;
    if (host !== "localhost" || !true) {
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-138890986-1');
    }
  </script>









  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript><!-- hexo-inject:begin --><!-- hexo-inject:end -->

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">singlecell.me</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    
      
    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>Home</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    
      
    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>Archives<span class="badge">4</span></a>

  </li>

      
      
    </ul>
  

  

  
</nav>



  



</div>
    </header>

    
  
  

  

  <a href="https://github.com/zhewa" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="80" height="80" viewbox="0 0 250 250" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a>



    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://zhewa.github.io/2019/05/01/processCellRangerOutputs/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Zhe Wang">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="singlecell.me">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">An R script to process Cell Ranger output files

              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-05-01 10:29:03" itemprop="dateCreated datePublished" datetime="2019-05-01T10:29:03-04:00">2019-05-01</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-05-05 12:27:08" itemprop="dateModified" datetime="2019-05-05T12:27:08-04:00">2019-05-05</time>
              
            
          </span>

          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2019/05/01/processCellRangerOutputs/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count disqus-comment-count" data-disqus-identifier="2019/05/01/processCellRangerOutputs/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2019/05/01/processCellRangerOutputs/" class="leancloud_visitors" data-flag-title="An R script to process Cell Ranger output files">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
            </span>
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/hint.css/2.4.1/hint.min.css"><p>First install the required packages <code>data.table</code>, <code>Matrix</code>, and <code>SingleCellExperiment</code>.<br><figure class="highlight objc"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">"data.table"</span>, quietly = <span class="literal">TRUE</span>)) &#123;</span><br><span class="line">    install.packages(<span class="string">"data.table"</span>)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">"Matrix"</span>, quietly = <span class="literal">TRUE</span>)) &#123;</span><br><span class="line">    install.packages(<span class="string">"Matrix"</span>)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">"SingleCellExperiment"</span>, quietly = <span class="literal">TRUE</span>)) &#123;</span><br><span class="line">    <span class="keyword">if</span> (!requireNamespace(<span class="string">"BiocManager"</span>, quietly = <span class="literal">TRUE</span>)) &#123;</span><br><span class="line">        install.packages(<span class="string">"BiocManager"</span>)</span><br><span class="line">    &#125;</span><br><span class="line">    BiocManager::install(<span class="string">"SingleCellExperiment"</span>)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure><a id="more"></a>The definition of function <code>processCellRangerOutput</code> is shown below. This function reads the count files (in sparse matrix format), the cell barcodes, and the feature annotations for each sample, writes each of the count matrices as a tab separated text file, then combines the count matrices as one for all samples, and finally writes it as a tab separated text file as well as a <code>SingleCellExperiment</code> object to specified path.</p>
<p>Note that this process requires a lot of memory depending on the size of the data. It might fail if the free memory on the computer is not enough for allocation or the data is too large to fit in the RAM.<br><figure class="highlight r"><figcaption><span>processCellRangerOutput.R</span><a href="https://github.com/zhewa/usefulscripts/blob/master/processCellRangerOutput.R" target="_blank" rel="noopener">link</a></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#' @title Process Cell Ranger output</span></span><br><span class="line"><span class="comment">#' @description Read the filtered matrices for all samples from Cell Ranger</span></span><br><span class="line"><span class="comment">#'  output. Combine them into one big expression matrix and write it to</span></span><br><span class="line"><span class="comment">#'  \code&#123;outDir&#125;. A \link[SingleCellExperiment]&#123;SingleCellExperiment&#125; object is</span></span><br><span class="line"><span class="comment">#'  also saved as a `.RDS` file.</span></span><br><span class="line"><span class="comment">#' @param cellRangerDir The root path where Cell Ranger was run. This folder</span></span><br><span class="line"><span class="comment">#'  should contain sample specific folders.</span></span><br><span class="line"><span class="comment">#' @param samples A vector of sample names. Must be the same as the folder names</span></span><br><span class="line"><span class="comment">#'  of the samples.</span></span><br><span class="line"><span class="comment">#' @param project The project name. Can be any string. It is</span></span><br><span class="line"><span class="comment">#'  added to the filenames of the combined count matrix and the</span></span><br><span class="line"><span class="comment">#'  \code&#123;SingleCellExperiment&#125; object.</span></span><br><span class="line"><span class="comment">#' @param writeIndivisualSampleCounts Boolean. Whether to write sample specific</span></span><br><span class="line"><span class="comment">#'  count matrices to output files or not. Default \code&#123;TRUE&#125;.</span></span><br><span class="line"><span class="comment">#' @param writeAllSampleCounts Boolean. Whether to write combined count matrix</span></span><br><span class="line"><span class="comment">#'  to output file or not. Default \code&#123;TRUE&#125;.</span></span><br><span class="line"><span class="comment">#' @param writeSCE Boolean. Whether to save the \code&#123;SingleCellExperiment&#125;</span></span><br><span class="line"><span class="comment">#'  object containing the combined count matrix to output file or not. Default</span></span><br><span class="line"><span class="comment">#'  \code&#123;TRUE&#125;.</span></span><br><span class="line"><span class="comment">#' @param sep The field separator string. Values within each row are separated</span></span><br><span class="line"><span class="comment">#'  by this string when writing output files.</span></span><br><span class="line"><span class="comment">#' @param outDir The path to output location. Default to current working</span></span><br><span class="line"><span class="comment">#'  directory.</span></span><br><span class="line"><span class="comment">#' @param outPrefix Prefix to output files. Default to current date.</span></span><br><span class="line"><span class="comment">#' @param overwrite Boolean. Whether to overwrite output files or not. Default</span></span><br><span class="line"><span class="comment">#'  \code&#123;FALSE&#125;.</span></span><br><span class="line"><span class="comment">#' @param cellRangerOuts The intermidiate path to filtered feature count files</span></span><br><span class="line"><span class="comment">#'  saved in sparce matrix format. Reference genome names might need to be</span></span><br><span class="line"><span class="comment">#'  appended if reads were mapped to multiple genomes when running Cell Ranger</span></span><br><span class="line"><span class="comment">#'  pipeline. Default \code&#123;"outs/filtered_feature_bc_matrix/"&#125;.</span></span><br><span class="line"><span class="comment">#' @param gzipped Boolean. Whether the Cell Ranger output files were gzip</span></span><br><span class="line"><span class="comment">#'  compressed or not. This is true subsequent to Cell Ranger 3.0. Default</span></span><br><span class="line"><span class="comment">#'  \code&#123;TRUE&#125;.</span></span><br><span class="line"><span class="comment">#' @return A \code&#123;SingleCellExperiment&#125; object containing the combined count</span></span><br><span class="line"><span class="comment">#'  matrix, the feature annotations, and the cell annotation.</span></span><br><span class="line">processCellRangerOutput &lt;- <span class="keyword">function</span>(</span><br><span class="line">    cellRangerDir,</span><br><span class="line">    samples,</span><br><span class="line">    project = strtrim(paste0(samples, collapse = <span class="string">""</span>), <span class="number">5</span>),</span><br><span class="line">    writeIndivisualSampleCounts = <span class="literal">TRUE</span>,</span><br><span class="line">    writeAllSampleCounts = <span class="literal">TRUE</span>,</span><br><span class="line">    writeSCE = <span class="literal">TRUE</span>,</span><br><span class="line">    sep = <span class="string">"\t"</span>,</span><br><span class="line">    outDir = <span class="string">"."</span>,</span><br><span class="line">    outPrefix = format(Sys.time(), <span class="string">"%Y%m%d"</span>),</span><br><span class="line">    overwrite = <span class="literal">FALSE</span>,</span><br><span class="line">    cellRangerOuts = <span class="string">"outs/filtered_feature_bc_matrix/"</span>,</span><br><span class="line">    gzipped = <span class="literal">TRUE</span>) &#123;</span><br><span class="line"></span><br><span class="line">    fileNames &lt;- c(matrix = <span class="string">"matrix.mtx"</span>,</span><br><span class="line">        features = <span class="string">"features.tsv"</span>,</span><br><span class="line">        barcodes = <span class="string">"barcodes.tsv"</span>)</span><br><span class="line">    <span class="keyword">if</span> (isTRUE(gzipped)) &#123;</span><br><span class="line">        fileNames &lt;- vapply(fileNames,</span><br><span class="line">            FUN = <span class="keyword">function</span>(i) &#123;</span><br><span class="line">                paste0(i, <span class="string">".gz"</span>)</span><br><span class="line">            &#125;,</span><br><span class="line">            FUN.VALUE = character(<span class="number">1</span>))</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    res &lt;- vector(<span class="string">"list"</span>, length = length(samples))</span><br><span class="line">    coldata &lt;- vector(<span class="string">"list"</span>, length = length(samples))</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> (i <span class="keyword">in</span> seq_along(samples)) &#123;</span><br><span class="line">        mm &lt;- Matrix::readMM(gzfile(file.path(cellRangerDir,</span><br><span class="line">            samples[i], cellRangerOuts, fileNames[[<span class="string">"matrix"</span>]])))</span><br><span class="line"></span><br><span class="line">        fe &lt;- read.table(gzfile(file.path(cellRangerDir,</span><br><span class="line">            samples[i], cellRangerOuts, fileNames[[<span class="string">"features"</span>]])), sep = <span class="string">"\t"</span>)</span><br><span class="line"></span><br><span class="line">        bc &lt;- read.table(gzfile(file.path(cellRangerDir,</span><br><span class="line">            samples[i], cellRangerOuts, fileNames[[<span class="string">"barcodes"</span>]])), sep = <span class="string">"\t"</span>)</span><br><span class="line">        coldata[[i]] &lt;- rep(samples[i], nrow(bc))</span><br><span class="line"></span><br><span class="line">        ma &lt;- as.matrix(mm)</span><br><span class="line">        colnames(ma) &lt;- paste(bc[[<span class="number">1</span>]], samples[i], sep = <span class="string">"_"</span>)</span><br><span class="line">        rownames(ma) &lt;- fe[[<span class="number">1</span>]]</span><br><span class="line">        res[[i]] &lt;- ma</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> (isTRUE(writeIndivisualSampleCounts)) &#123;</span><br><span class="line">            fn &lt;- file.path(outDir,</span><br><span class="line">                paste0(outPrefix, <span class="string">"_"</span>, samples[i], <span class="string">".txt"</span>))</span><br><span class="line">            <span class="keyword">if</span> (isFALSE(overwrite) &amp; file.exists(fn)) &#123;</span><br><span class="line">                <span class="keyword">warning</span>(<span class="string">"File already exists! Skip writting file "</span>, fn)</span><br><span class="line">            &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">                data.table::fwrite(ma,</span><br><span class="line">                    file = fn,</span><br><span class="line">                    sep = sep,</span><br><span class="line">                    row.names = <span class="literal">TRUE</span>)</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    expr &lt;- do.call(cbind, res)</span><br><span class="line">    exprdt &lt;- data.table::as.data.table(expr, keep.rownames = <span class="literal">TRUE</span>)</span><br><span class="line">    colnames(exprdt)[<span class="number">1</span>] &lt;- <span class="string">"geneID"</span></span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> (isTRUE(writeAllSampleCounts)) &#123;</span><br><span class="line">        fn &lt;- file.path(outDir,</span><br><span class="line">            paste0(outPrefix, <span class="string">"_"</span>, project, <span class="string">"_allcounts.txt"</span>))</span><br><span class="line">        <span class="keyword">if</span> (isFALSE(overwrite) &amp; file.exists(fn)) &#123;</span><br><span class="line">            <span class="keyword">warning</span>(<span class="string">"File already exists! Skip writting file "</span>, fn)</span><br><span class="line">        &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">            data.table::fwrite(exprdt,</span><br><span class="line">                file = fn,</span><br><span class="line">                sep = sep,</span><br><span class="line">                row.names = <span class="literal">TRUE</span>)</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    sce &lt;- SingleCellExperiment::SingleCellExperiment(</span><br><span class="line">        assays = list(counts = expr))</span><br><span class="line">    SummarizedExperiment::rowData(sce) &lt;- fe</span><br><span class="line">    colnames(SummarizedExperiment::rowData(sce)) &lt;- c(<span class="string">"geneID"</span>,</span><br><span class="line">        <span class="string">"genename"</span>, <span class="string">"genebiotype"</span>)</span><br><span class="line">    coldata &lt;- unlist(coldata)</span><br><span class="line">    SummarizedExperiment::colData(sce)[<span class="string">"sample"</span>] &lt;- data.frame(sample = coldata)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> (isTRUE(writeSCE)) &#123;</span><br><span class="line">        fn &lt;- file.path(outDir,</span><br><span class="line">            paste0(outPrefix, <span class="string">"_"</span>, project, <span class="string">"_allcounts_sce.RDS"</span>))</span><br><span class="line">        <span class="keyword">if</span> (isFALSE(overwrite) &amp; file.exists(fn)) &#123;</span><br><span class="line">            <span class="keyword">warning</span>(<span class="string">"File already exists! Skip writting file "</span>, fn)</span><br><span class="line">        &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">            saveRDS(sce, file = fn)</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span>(sce)</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure></p>

      
    </div>

    

    
    
    

    

    
      
    
    

    
      <div>
        




  



<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Zhe Wang</li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    
    <a href="https://zhewa.github.io/2019/05/01/processCellRangerOutputs/" title="An R script to process Cell Ranger output files">https://zhewa.github.io/2019/05/01/processCellRangerOutputs/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fa fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.</li>
</ul>

      </div>
    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Single-Cell/" rel="tag"># Single Cell</a>
          
            <a href="/tags/Script/" rel="tag"># Script</a>
          
            <a href="/tags/Cell-Ranger/" rel="tag"># Cell Ranger</a>
          
            <a href="/tags/R/" rel="tag"># R</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/04/22/peerreviewethics/" rel="next" title="The Ethical Issue of Pre-Publication Single Blind Peer Review">
                <i class="fa fa-chevron-left"></i> The Ethical Issue of Pre-Publication Single Blind Peer Review
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/05/05/scruff/" rel="prev" title="scruff, a new R/Bioconductor package for preprocessing scRNA-seq data">
                scruff, a new R/Bioconductor package for preprocessing scRNA-seq data <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>Please enable JavaScript to view the comments powered by Disqus.</noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      

      <div class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">Zhe Wang</p>
              <div class="site-description motion-element" itemprop="description"></div>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">4</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              

              
                
                
                <div class="site-state-item site-state-tags">
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">10</span>
                    <span class="site-state-item-name">tags</span>
                  
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          

          

          
             <div class="cc-license motion-element" itemprop="license">
              
              
                
              
              
              
              <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" class="cc-opacity" rel="noopener" target="_blank"><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"></a>
             </div>
          

          
          

          
            
          
          

        </div>
      </div>

      

      

    </div>
  </aside>
  


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2019 – <span itemprop="copyrightYear">2020</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Zhe Wang</span>

  
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-area-chart"></i>
    </span>
    
    <span title="Symbols count total">23k</span>
  

  
</div>


  <div class="powered-by">Powered by <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> v3.8.0</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a> v7.1.0</div>




        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="post-meta-item-icon">
      <i class="fa fa-user"></i>
    </span>
    <span class="site-uv" title="Total Visitors">
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
    </span>
  

  
    <span class="post-meta-divider">|</span>
  

  
    <span class="post-meta-item-icon">
      <i class="fa fa-eye"></i>
    </span>
    <span class="site-pv" title="Total Views">
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>


  


  <script src="/js/utils.js?v=7.1.0"></script>

  <script src="/js/motion.js?v=7.1.0"></script>



  
  


  <script src="/js/affix.js?v=7.1.0"></script>

  <script src="/js/schemes/pisces.js?v=7.1.0"></script>




  
  <script src="/js/scrollspy.js?v=7.1.0"></script>
<script src="/js/post-details.js?v=7.1.0"></script>



  


  <script src="/js/next-boot.js?v=7.1.0"></script>


  

  

  

  
  
  <script id="dsq-count-scr" src="https://singlecell.disqus.com/count.js" async></script>


<script>
  var disqus_config = function() {
    this.page.url = "https://zhewa.github.io/2019/05/01/processCellRangerOutputs/";
    this.page.identifier = "2019/05/01/processCellRangerOutputs/";
    this.page.title = 'An R script to process Cell Ranger output files';
    };
  function loadComments() {
    var d = document, s = d.createElement('script');
    s.src = 'https://singlecell.disqus.com/embed.js';
    s.setAttribute('data-timestamp', '' + +new Date());
    (d.head || d.body).appendChild(s);
  }
  
    $(function() {
      var offsetTop = $('#comments').offset().top - $(window).height();
      if (offsetTop <= 0) {
        // load directly when there's no a scrollbar
        loadComments();
      } else {
        $(window).on('scroll.disqus_scroll', function() {
          // offsetTop may changes because of manually resizing browser window or lazy loading images.
          var offsetTop = $('#comments').offset().top - $(window).height();
          var scrollTop = $(window).scrollTop();

          // pre-load comments a bit? (margin or anything else)
          if (offsetTop - scrollTop < 60) {
            $(window).off('.disqus_scroll');
            loadComments();
          }
        });
      }
    });
  
</script>





  


  




  
  
  <script>
    
    function addCount(Counter) {
      var $visitors = $('.leancloud_visitors');
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();

      Counter('get', '/classes/Counter', { where: JSON.stringify({ url }) })
        .done(function({ results }) {
          if (results.length > 0) {
            var counter = results[0];
            
            Counter('put', '/classes/Counter/' + counter.objectId, JSON.stringify({ time: { '__op': 'Increment', 'amount': 1 } }))
            
              .done(function() {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.time + 1);
              })
            
              .fail(function ({ responseJSON }) {
                console.log('Failed to save Visitor num, with error message: ' + responseJSON.error);
              })
          } else {
            
              Counter('post', '/classes/Counter', JSON.stringify({ title: title, url: url, time: 1 }))
                .done(function() {
                  var $element = $(document.getElementById(url));
                  $element.find('.leancloud-visitors-count').text(1);
                })
                .fail(function() {
                  console.log('Failed to create');
                });
            
          }
        })
        .fail(function ({ responseJSON }) {
          console.log('LeanCloud Counter Error: ' + responseJSON.code + ' ' + responseJSON.error);
        });
    }
    

    $(function() {
      $.get('https://app-router.leancloud.cn/2/route?appId=' + 'aIrtJIi2bwNvltN6njsaeRvY-MdYXbMMI')
        .done(function({ api_server }) {
          var Counter = function(method, url, data) {
            return $.ajax({
              method: method,
              url: 'https://' + api_server + '/1.1' + url,
              headers: {
                'X-LC-Id': 'aIrtJIi2bwNvltN6njsaeRvY-MdYXbMMI',
                'X-LC-Key': 'iV6yV65mhvSG4eUuyXstOJwa',
                'Content-Type': 'application/json',
              },
              data: data
            });
          };
          
            addCount(Counter);
          
        });
    });
  </script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({"tex2jax":{"inlineMath":[["$","$"],["\\(","\\)"]],"skipTags":["script","noscript","style","textarea","pre","code"],"processEscapes":true},"TeX":{"equationNumbers":{"autoNumber":"AMS"}}});
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->



  

  
  

  
  

  


  

  

  

  

  

  

  

  

  

  

  

</body>
</html>
